src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_NullCoalescingOperator.cs

```diff
@@ -132,13 +130,27 @@ public override BoundNode VisitNullCoalescingOperator(BoundNullCoalescingOperato
                 }
             }
 
-            // Optimize left ?? right to left.GetValueOrDefault() when left is T? and right is the default value of T
-            if (rewrittenLeft.Type.IsNullableType()
-                && RemoveIdentityConversions(rewrittenRight).IsDefaultValue()
-                && rewrittenRight.Type.Equals(rewrittenLeft.Type.GetNullableUnderlyingType(), TypeCompareKind.AllIgnoreOptions)
-                && TryGetNullableMethod(rewrittenLeft.Syntax, rewrittenLeft.Type, SpecialMember.System_Nullable_T_GetValueOrDefault, out MethodSymbol getValueOrDefault))
+            // Optimizations related to nullable value types
+            if (rewrittenLeft.Type.IsNullableType())
             {
-                return BoundCall.Synthesized(rewrittenLeft.Syntax, rewrittenLeft, initialBindingReceiverIsSubjectToCloning: ThreeState.Unknown, getValueOrDefault);
+                var leftUnderlyingType = rewrittenLeft.Type.GetNullableUnderlyingType();
+                var unwrappedRight = RemoveIdentityConversions(rewrittenRight);
+
+                // Optimize left ?? right to left.GetValueOrDefault() when left is T? and right is the default value of T
+                if (unwrappedRight.IsDefaultValue() &&
+                    rewrittenRight.Type.Equals(leftUnderlyingType, TypeCompareKind.AllIgnoreOptions) &&
+                    TryGetNullableMethodSoft(rewrittenLeft.Type, SpecialMember.System_Nullable_T_GetValueOrDefault, out MethodSymbol? getValueOrDefault))
+                {
+                    return BoundCall.Synthesized(rewrittenLeft.Syntax, rewrittenLeft, initialBindingReceiverIsSubjectToCloning: ThreeState.Unknown, getValueOrDefault);
+                }
+
+                // Optimize left ?? right to left.GetValueOrDefault(right) when left is T? and right value is a constant of type T
+                if (unwrappedRight.ConstantValueOpt is not null &&
+                    rewrittenRight.Type.Equals(leftUnderlyingType, TypeCompareKind.AllIgnoreOptions) &&
```

Comment:
Perhaps extract
```csharp
rewrittenRight.Type.Equals(leftUnderlyingType, TypeCompareKind.AllIgnoreOptions)
```
in an outer if, but also preserve evaluation of `IsDefaultValue` and `ConstantValueOpt is not null` first?